docs: add troubleshooting page (#9379)
vitejs/vite: docs/guide/troubleshooting.md

# Troubleshooting

> See [rollup's troubleshooting](https://rollupjs.org/guide/en/#troubleshooting) too.

Reviewer: ```suggestion
See [Rollup's troubleshooting](https://rollupjs.org/guide/en/#troubleshooting) guide for more information too.
```

> If you didn't make it work, please try posting questions on [GitHub Discussions](https://github.com/vitejs/vite/discussions) or discussioning the issue in #help channel on [Vite Land Discord](https://chat.vitejs.dev).

Reviewer: ```suggestion
If the suggestions here don't work, please try posting questions on [GitHub Discussions](https://github.com/vitejs/vite/discussions) or in the `#help` channel of [Vite Land Discord](https://chat.vitejs.dev).
```

## CLI

### `Error: Cannot find module 'C:\foo\bar&baz\vite\bin\vite.js'`

> Probably the path to your project folder includes `?`.
> There is a bug in `npm` when using Windows and the path including `?` ([npm/cmd-shim#45](https://github.com/npm/cmd-shim/issues/45)).

Reviewer: ```suggestion
The path to your project folder may include `?`, which doesn't work with `npm` on Windows ([npm/cmd-shim#45](https://github.com/npm/cmd-shim/issues/45)).
```

Simplified this part a bit.

I think we should also move single paragraphs all in the same line as it's common for markdown files to line wrap in the IDE.

You will need to either:

- Switch to other package manager (e.g. `pnpm`, `yarn`)
- Remove `?` from the path to your project

## Dev Server

### Requests are stalled forever

> If you are using linux, file descriptor limits and inotify limits may be affecting.
> Because Vite does not bundle most of the files, browsers requests many files and this requires many file descriptors.

Reviewer: ```suggestion
If you are using Linux, file descriptor limits and inotify limits may be causing the issue. As Vite does not bundle most of the files, browsers may request many files which require many file descriptors, going over the limit.
```

To solve this:

- Increase file descriptor limits by `ulimit`
- Increase the following inotify related limits by `sysctl`
  - `fs.inotify.max_queued_events`
  - `fs.inotify.max_user_instances`

>   - `fs.inotify.max_user_watches`

Reviewer: I think we could use some code examples here. I think https://github.com/vitejs/vite/issues/5310#issuecomment-986042734 and https://github.com/vitejs/vite/issues/5310#issuecomment-1181860523 would do. Something like:

```bash
# Check current ulimit
$ ulimit -Sn

# Change limit
$ ulimit -Sn 10000
```

Make sure to restart your browser after changing the limit.

---

```bash
# Check current fs.inotify.* values
$ sysctl fs.inotify

# Change a fs.inotify.* value
$ sudo sysctl fs.inotify.max_user_instances=8192
```

Author: Done 👍

## HMR

### Vite detects a file change but the HMR is not working

Probably you are importing a file with different case.
For example, `src/foo.js` exists and `src/bar.js` contains:

```js
import './Foo.js' // should be './foo.js'
```

Related issue: [#964](https://github.com/vitejs/vite/issues/964)

### Vite does not detect a file change

> If you are running Vite with WSL2, Vite cannot watch file changes in some conditions.
> See [`server.watch` option](/config/server-options.md#server-watch).

Reviewer: ```suggestion
If you are running Vite with WSL2, Vite cannot watch file changes in some conditions. See [`server.watch` option](/config/server-options.md#server-watch).
```

# Troubleshooting

See [Rollup's troubleshooting guide](https://rollupjs.org/guide/en/#troubleshooting) for more information too.

If the suggestions here don't work, please try posting questions on [GitHub Discussions](https://github.com/vitejs/vite/discussions) or in the `#help` channel of [Vite Land Discord](https://chat.vitejs.dev).

## CLI

### `Error: Cannot find module 'C:\foo\bar&baz\vite\bin\vite.js'`

The path to your project folder may include `?`, which doesn't work with `npm` on Windows ([npm/cmd-shim#45](https://github.com/npm/cmd-shim/issues/45)).

You will need to either:

> - Switch to other package manager (e.g. `pnpm`, `yarn`)

Reviewer: ```suggestion
- Switch to another package manager (e.g. `pnpm`, `yarn`)
```

- Remove `?` from the path to your project

## Dev Server

### Requests are stalled forever

If you are using Linux, file descriptor limits and inotify limits may be causing the issue. As Vite does not bundle most of the files, browsers may request many files which require many file descriptors, going over the limit.

To solve this:

- Increase file descriptor limit by `ulimit`

  ```shell
  # Check current limit
  $ ulimit -Sn
  # Change limit (temporary)
  $ ulimit -Sn 10000 # You might need to change the hard limit too
  # Restart your browser
  ```

- Increase the following inotify related limits by `sysctl`

  ```shell
  # Check current limits
  $ sysctl fs.inotify
  # Change limits (temporary)
  $ sudo sysctl fs.inotify.max_queued_events=16384
  $ sudo sysctl fs.inotify.max_user_instances=8192
  $ sudo sysctl fs.inotify.max_user_watches=524288
  ```

## HMR

### Vite detects a file change but the HMR is not working

> Probably you are importing a file with different case.
> For example, `src/foo.js` exists and `src/bar.js` contains:

Reviewer: ```suggestion
You may be importing a file with a different case. For example, `src/foo.js` exists and `src/bar.js` contains:
```
